### PR TITLE
Fix #4031 - Adds background playback and pip to all web-pages

### LINKED ISSUE #4031: Add Background Playback and PIP for WebPages
## Feature Overview
- Websites like youtube do not support Background play or Picture-In-Picture.
- Want to be able to play videos (background audio) on lock screen and picture in picture when on the home-screen.

## Implementation Details
- Picture-In-Picture needs the `webkitpresentationmodechanged` event blocked from propagating/bubbling.
- Background-Mode is trickier as Webkit itself calls `pause` on the `HTMLMediaElement` so you need to `resume` playback. However, websites like `youtube` also break playback even if it were fixed in youtube, by listening to the `visibilitychange` event and pausing the `HTMLMediaElement`.

## User Experience 

1. Play a video on a website
2. Background the browser
3. Should still play.
4. Lock the device.
5. Should still play.
6. Pause and then play again on the lock-screen; should work.

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary of Changes
- Security Review: https://github.com/brave/security/issues/550
- Injects JS that forces media to continue playing when the page is backgrounded for websites that purposely break background playback which is an OS feature.
- Injects JS that stops propagating the `webkitpresentationmodechanged` event for PIP support on websites that purposely break PIP which is supported OS wide.

This pull request fixes #4031

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

**Review by iccub — COMMENTED** [review]
Looks fine to me, two questions;
1. Should we make it a settings toggle? Any chance this may break webcompat on websites?
2. I would add a security review for it

File: Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+          mutations.addedNodes.forEach(function (node) {
+              if (node.constructor.name == 'HTMLVideoElement') {
+                  addListeners(node);
+              }
+          });
+      }
+      queue.length = 0;
+    }
+    
+    var observer = new MutationObserver(function(mutations) {
```
Comment (diracdeltas) [inline Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js:107]: cc @kylehickinson; i know there have been concerns in the past around performance with MutationObserver

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 2]
```diff
@@ -304,6 +304,27 @@ class UserScriptManager {
...
+    private let MediaBackgroundingScript: WKUserScript? = {
+        guard let path = Bundle.main.path(forResource: "MediaBackgrounding", ofType: "js"), let source = try? String(contentsOfFile: path) else {
+            log.error("Failed to load MediaBackgrounding.js")
+            return nil
+        }
+        
+        var alteredSource = source
+        let token = UserScriptManager.securityToken.uuidString.replacingOccurrences(of: "-", with: "", options: .literal)
+        
+        let replacements = [
```
Comment (diracdeltas) [inline Client/Frontend/Browser/UserScriptManager.swift:331]: what is the purpose of doing these replacements? cc @jumde

Comment (Brandon-T) [reply Client/Frontend/Browser/UserScriptManager.swift:331]: When you override a property or inject code, someone's page can then override or replace that function with its own implementation. So the solution is to randomly generate the name and make it non-enumerable so the website cannot see it or change it.

This is solved in iOS 14+ with `WKContentWorld` but we currently support iOS 13+.

File: Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    };
+}
+
+
+window.__firefox__.includeOnce("$<MediaBackgrounding>", function() {
+    var visibilityState_Get = Object.getOwnPropertyDescriptor(Document.prototype, "visibilityState").get;
+    var visibilityState_Set = Object.getOwnPropertyDescriptor(Document.prototype, "visibilityState").set;
+    Object.defineProperty(Document.prototype, 'visibilityState', {
+        enumerable: true,
+        configurable: true,
```
Comment (jumde) [inline Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js:8]: nit: Setting `configurable` to true allows scripts to delete this property. Do we need this property to be configurable?

File: Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    };
+}
+
+
+window.__firefox__.includeOnce("$<MediaBackgrounding>", function() {
+    var visibilityState_Get = Object.getOwnPropertyDescriptor(Document.prototype, "visibilityState").get;
+    var visibilityState_Set = Object.getOwnPropertyDescriptor(Document.prototype, "visibilityState").set;
+    Object.defineProperty(Document.prototype, 'visibilityState', {
+        enumerable: true,
```
Comment (jumde) [inline Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js:7]: @Brandon-T - `window.__firefox__.userScripts` reveals the secret token.

File: Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
+// The below is needed because the script may not be web-packed into a bundle so it may be missing the run-once code
+
+// MARK: - Include Once
+
+if (!window.__firefox__) {
```
Comment (jumde) [inline Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js]: @Brandon-T - This needs to be removed right?

File: Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js (hunk 3 above)
Comment (Brandon-T) [reply Client/Frontend/UserContent/UserScripts/MediaBackgrounding.js:8]: This property is by default configurable as it is a property on `HTMLDocument` itself.

**Review by jumde — APPROVED** [review]
lgtm

**Pull request merged**